### DISCUSSION TITLE
ROSCO Controller Speed Exclusion Zone

### DISCUSSION THREAD

**Discussion by Sam-Ramsahoye** [NatLabRockies/ROSCO#207] [Q&A] [answered]
Hello,

I'd like to be able to avoid the tower natural frequency in region 2 for a fixed-base turbine. It looks as if some work has been done relating to this and I can see in Tune_Cases there are files like IEA15MW_MultiOmega.yaml and NREL5MW_PassThrough.yaml. 

If it's possible to define an exclusion zone of rotational speeds in region 2, it would be great if someone could explain how I might setup the appropriate input files.

Moreover, it would be great if someone could very briefly explain how the ROSCO speed exclusion zone works as I'm aware there are a number of different possible implementations.

![MicrosoftTeams-image](https://user-images.githubusercontent.com/50165138/210544650-b0fc30dd-c9d1-4bfb-850e-74bef0302be2.png)

Best,
Sam

**Comment by dzalkind** [comment, marked as answer]
Hi Sam,

The draft PR of the ROSCO exclusion zone works by changing the reference generator speed that is tracked by the torque controller: https://github.com/dzalkind/ROSCO/blob/16bd05c66644b31d629920c915bdad827cbb222d/ROSCO/src/ControllerBlocks.f90#L57 

If the generator speed is below the exclusion speed, the reference speed is held at the exclusion speed minus the exclusion band, and vice versa: https://github.com/dzalkind/ROSCO/blob/16bd05c66644b31d629920c915bdad827cbb222d/ROSCO/src/ControllerBlocks.f90#L445 

The main issue that we haven't had time to work out is that the generator speed tracking is poor by design to avoid large generator torque (and power) transients.  I think we need to increase the torque control gains when in or near the exclusion region and ideally do so smoothly.  

The branch that I linked to is a little out of date.  I can see about bringing it up to date with the latest `develop` branch.  If you're interested in further developing and testing these features, I would be happy to provide support.

Best, Dan

**Reply by Sam-Ramsahoye** [reply]
@dzalkind Thanks very much for your reply. I'll take a look at the source you've linked and give it a go. I see you added some new commits yesterday too. I'm certainly interested in testing and developing out this feature - I'll keep you updated if I make any progress or if I need some guidance.

Cheers,
Sam

**Comment by dzalkind** [comment]
I just made another push to this branch, including `Twr_GainFactor` and `Twr_GainTau` as inputs to ROSCO in order to address the main issues I had with the algorithm.  I have not settled on "good" values for each of these, but there is an example in that branch now, as well: https://github.com/dzalkind/ROSCO/blob/f/exclusion_zone/Examples/example_19.py 

<img width="407" alt="image" src="https://user-images.githubusercontent.com/65573423/211905572-2a514abc-1328-4160-9cea-683fe3a2ba50.png">

**Comment by Sam-Ramsahoye** [comment]
@dzalkind Thanks very much for providing the example, it's very helpful and looks really promising!

I was able to get the example you pushed running and I can match the plots you've provided. Are the large transients in the torque partly a consequence of such a large speed exclusion zone, and so a large change in set point when traversing it?

Forgive me as I'm no control expert, nor am I as familiar with ROSCO as you are, but could the transients be alleviated in anyway by filtering the reference speed so that it doesn't snap as quickly across the speed exclusion zone? I have had a look at the source code and I don't think this is being done, although I may be mistaken and perhaps it's not necessary either.

I'm going to be away over the next two weeks but I'm going to try and find some time to experiment with this further over my break.

Cheers,
Sam

**Reply by dzalkind** [reply]
Hi Sam,

The transients are because of the sharp changes in the set point, which you could filter.  I would be curious to see how that looks.  

I am also increasing the gains near the hysteresis region, which makes the transients worse than if the gains were constant. [Here](https://github.com/dzalkind/ROSCO/blob/b3a8b1cbf0c59c93ccaa6e3a8abbf06d80b7600e/ROSCO/src/ControllerBlocks.f90#L484) is where I filter the gains if you'd like to use that as a template.  I think I would prefer the filter to have a slower "take-off" than it currently has. 

Please share anything promising you find, and we can incorporate it.

Best, Dan

<img width="555" alt="image" src="https://user-images.githubusercontent.com/65573423/212116368-de41ddce-9a49-4cba-a230-e1cf6542eeb7.png">

**Reply by Sam-Ramsahoye** [reply]
@dzalkind Thanks Dan - I'm looking forward to have some time to dig into it properly.
